### PR TITLE
feat: add RL synthetic fixtures, reference ops, and minimal loss-step tests

### LINKED ISSUE #27: [FEAT][alignment]: add minimal RL batch schema and synthetic GRPO fixtures for kernel validation
## Summary

Add a generic RL-shaped batch schema and deterministic synthetic fixtures for validating Kernel-Align kernels without requiring a full RL trainer, reward model, vLLM worker, DeepSpeed worker, or Ray cluster.

`fused_logp` should be the first consumer, but the fixture must be generic enough for future sampling, attention, masked reduction, and RL loss kernels.

## Why This Issue Exists

Kernel-Align is an RL post-training infrastructure project, but kernels currently have only narrow random-tensor tests. That misses key RL workload properties:

- prompt tokens and completion tokens have different semantics;
- padding and ignored tokens must be masked;
- group size `G` affects memory pressure;
- completion lengths vary across samples;
- current, old, and reference policy quantities are compared;
- rewards and advantages feed loss kernels;
- future kernels may need sequence metadata, not just logits/token ids.

A generic synthetic RL batch fixture lets every kernel test speak the same workload language.

## Scope

Create a lightweight fixture module, for example:

- `rl_engine/testing/rl_batch.py`, preferred if benchmarks and tests will both use it; or
- `tests/fixtures/rl_batch.py`, if it should stay test-only at first.

The fixture should support tensors and metadata such as:

- `input_ids`: `[batch_or_group, total_seq_len]`
- `position_ids`: optional `[batch_or_group, total_seq_len]`
- `attention_mask`: `[batch_or_group, total_seq_len]`
- `prompt_mask`: `[batch_or_group, total_seq_len]`
- `completion_mask`: `[batch_or_group, total_seq_len]` or `[batch_or_group, completion_len]`
- `token_ids`: selected next-token ids for logprob kernels
- `rewards`: synthetic scalar or per-token rewards
- `advantages`: `[batch_or_group, completion_len]`
- `old_logps`: `[batch_or_group, completion_len]`
- `ref_logps`: `[batch_or_group, completion_len]`
- optional `temperature`
- optional `row_indices` / `valid_indices` for compact valid-token kernel paths

Configurable dimensions:

- batch size or group size `G`;
- number of prompts;
- samples per prompt;
- prompt length;
- completion length;
- vocab size;
- hidden size, optional for future LM-head or attention tests;
- number of heads/head dimension, optional for attention tests;
- dtype;
- device;
- valid-token mask density;
- deterministic seed.

## Non-Goals

- Do not implement a real reward model.
- Do not implement a full GRPO/PPO/DPO trainer.
- Do not integrate vLLM, DeepSpeed, Ray, TRL, OpenRLHF, or verl.
- Do not optimize CUDA kernels in this issue.

## Suggested Implementation

Add a dataclass such as:

```python
@dataclass
class SyntheticRLKernelBatch:
    input_ids: torch.Tensor
    attention_mask: torch.Tensor
    prompt_mask: torch.Tensor
    completion_mask: torch.Tensor
    token_ids: torch.Tensor
    rewards: torch.Tensor
    advantages: torch.Tensor
    old_logps: torch.Tensor
    ref_logps: torch.Tensor
    valid_indices: torch.Tensor | None = None
    metadata: dict[str, Any] = field(default_factory=dict)
```

Add a generator such as:

```python
batch = make_synthetic_rl_kernel_batch(
    num_prompts=4,
    samples_per_prompt=8,
    prompt_len=32,
    completion_len=128,
    vocab_size=4096,
    valid_density=0.75,
    dtype=torch.float16,
    device="cuda",
    seed=0,
)
```

Add helper views:

- dense completion view;
- flattened valid-token view;
- compact row indices for kernels that skip masked tokens;
- metadata summary for benchmark CSV rows.

## Acceptance Criteria

- Tests can create deterministic RL-shaped batches on CPU and CUDA.
- Fixture supports prompt/completion/padding masks.
- Fixture supports group-style generation with multiple samples per prompt.
- Fixture can produce valid-token row indices for future compact kernels.
- Fixture does not require external model weights, datasets, or services.
- At least one unit test verifies shapes, masks, deterministic generation, and valid-token indexing.
- `fused_logp` can consume the fixture as the first example, without making the fixture logprob-only.

## Dependencies

- None.

## Follow-Up Users

- RL kernel reference operator issue.
- RL-shaped kernel benchmark issue.
- Minimal RL loss-step kernel test issue.
- `fused_logp_kernel.cu` optimization PR.
- Future sampling, attention, and loss kernel PRs.

### LINKED ISSUE #28: [FEAT][alignment]: implement selected-logprob reference path for current/old/ref policy testing
## Summary

Add PyTorch reference operators that define the semantics expected from Kernel-Align RL kernels. These references should be slow but clear, deterministic, and reusable across kernel tests.

`selected_logprobs_reference` should be the first reference because issue #5 needs it, but the module should be organized as a generic RL kernel reference layer.

## Why This Issue Exists

CUDA kernels need a precise semantic contract before they are optimized. In RL post-training, small numerical or masking errors can affect:

- policy ratio;
- KL penalty;
- clipping behavior;
- active-token normalization;
- sampling/training logprob consistency;
- reward or advantage weighting.

A generic reference operator layer lets kernel tests compare against one trusted source instead of duplicating ad hoc PyTorch snippets.

## Scope

Create a reference module, for example:

- `rl_engine/testing/reference_ops.py`, preferred for reuse by tests and benchmarks; or
- `tests/reference/rl_ops.py`, if it should stay test-only at first.

Initial reference operators:

- selected-token logprobs:
  - `log_softmax(logits.float(), dim=-1).gather(...)`
  - optional completion mask;
  - optional temperature;
  - explicit output dtype, with fp32 preferred for RL metrics.
- masked reductions:
  - masked mean;
  - masked sum;
  - active-token count.
- GRPO/PPO-style metrics:
  - `ratio = exp(current_logps - old_logps)`
  - KL approximation against reference logps;
  - ratio drift / KL drift helpers.
- optional sampling reference hooks:
  - top-k/top-p semantics can be added later only when the trainer contract defines expected behavior.

## Non-Goals

- Do not optimize CUDA code.
- Do not implement a full trainer.
- Do not add external RL framework dependencies.
- Do not define production top-k/top-p filtered logprob semantics until rollout/training contracts require it.

## Suggested Implementation

Keep references explicit and readable:

```python
def selected_logprobs_reference(
    logits: torch.Tensor,
    token_ids: torch.Tensor,
    mask: torch.Tensor | None = None,
    temperature: float = 1.0,
    output_dtype: torch.dtype = torch.float32,
) -> torch.Tensor:
    scaled_logits = logits.float() / temperature
    log_probs = torch.log_softmax(scaled_logits, dim=-1)
    selected = torch.gather(log_probs, dim=-1, index=token_ids.unsqueeze(-1)).squeeze(-1)
    if mask is not None:
        selected = selected.masked_fill(~mask, 0.0)
    return selected.to(output_dtype)
```

Add helpers such as:

```python
def masked_mean(values, mask, eps=1e-8): ...
def compute_policy_ratio(current_logps, old_logps, mask=None): ...
def compute_reference_kl(current_logps, ref_logps, mask=None): ...
def summarize_kernel_drift(candidate, reference, mask=None): ...
```

## Acceptance Criteria

- Reference tests cover fp32, fp16, and bfloat16 logits where supported.
- Reference tests cover dense and masked completion tokens.
- Reference tests cover temperature values such as `1.0` and `0.7`.
- Helpers report logprob error, ratio drift, KL drift, and active-token count.
- `fused_logp` tests can import the selected-logprob reference.
- The module is named and organized so future kernels can add reference functions without creating a logprob-only dependency.

## Dependencies

- Generic RL kernel batch schema and fixtures.

## Follow-Up Users

- RL-shaped kernel benchmark suite.
- Minimal RL loss-step kernel test.
- Native fused logp CUDA optimization candidates.
- Future sampling, attention, masked reduction, and loss kernel tests.

### LINKED ISSUE #30: [FEAT][tests]: add minimal GRPO loss-step test around fused_logp
Add a small synthetic GRPO loss-step test that validates fused logprob behavior in the context where the kernel will actually be used.

Scope:
- Use synthetic logits/token ids/advantages/old logps/ref logps.
- Compute selected current logps with:
  - PyTorch reference
  - fused logp kernel
- Compute a minimal GRPO/PPO-style clipped objective.
- Compare:
  - loss value
  - ratio
  - KL metric
  - active-token normalization
- No optimizer step or real model required.

Acceptance:
- Test fails if fused logp changes GRPO loss beyond tolerance.
- Test covers masked completions.
- Test documents whether logprobs are fp32 or low precision.

### PR DESCRIPTION
**Pull request by inaniloquentee**
## Summary

This PR adds the minimal RL-shaped validation stack needed to test kernel behavior without a full trainer or serving runtime.

It includes:

- reusable synthetic RL batch fixtures
- explicit PyTorch reference operators for selected logprobs, masked reductions, ratio, KL, and drift summaries
- minimal RL loss-step integration tests that validate kernel output inside objective math

## What changed

- Added 
l_engine.testing helpers for RL-shaped synthetic batches and reference ops.
- Added fixture coverage for determinism, masks, compact valid-token indexing, CUDA smoke, and grouped prompt sampling layout.
- Added minimal loss-step tests that verify masked tokens do not leak into ratio, KL, or loss computation.
- The CUDA fused-logp candidate path now skips cleanly when the fused backend is unavailable instead of failing through a generic fallback.

## Notes

- Grouped samples now share prompt tokens within each prompt group while keeping completions independent.
- The fixture and reference APIs are intentionally generic so future RL kernel tests and benchmarks can reuse them.

## Validation

`powershell
py -3.13 -m pytest tests/test_rl_batch_fixture.py tests/test_reference_ops.py tests/test_rl_kernel_loss_step.py -q
py -3.13 -m pytest tests -q
py -3.13 -m compileall rl_engine tests benchmarks
pre-commit run --files rl_engine/testing/__init__.py rl_engine/testing/rl_batch.py rl_engine/testing/reference_ops.py tests/conftest.py tests/test_rl_batch_fixture.py tests/test_reference_ops.py tests/test_rl_kernel_loss_step.py --show-diff-on-failure
`

## Issues

Closes #27  
Closes #28  
Closes #30

### CODE REVIEW

**Comment by Flink-ddd** [comment]
LGTM, Thank you for contribution.
